### ISSUE TITLE
[RFC] Add Gemma 3 Models

### ISSUE THREAD

**Issue by hashkanna** [sgl-project/sglang-jax#1069]
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

## 1. Abstract

Add native Gemma 3 support to sglang-jax in reviewable slices:

1. A Gemma 3 text decoder (`Gemma3ForCausalLM`) covering the 270M and 1B
   checkpoints.
2. A Gemma 3 vision path (`Gemma3ForConditionalGeneration`) covering the 4B,
   12B, and 27B checkpoints through the existing staged multimodal runtime
   from #552.

Gemma 3n is no longer part of this RFC. It is split into a separate follow-up
RFC #1078 because its PLE, AltUp/Laurel, shared-KV, audio, and MobileNet/TIMM tower
decisions are mostly independent from core Gemma 3 text and vision.

## 2. Motivation

#816 requests Gemma3-Vision support, and the sglang-jax Q2 roadmap (#909) lists
Gemma 3 under multimodal model targets. The repo already recognizes
`Gemma3ForConditionalGeneration` as a multimodal architecture name in
`python/sgl_jax/srt/configs/model_config.py`. Native model classes, weight mappings, a SigLIP
vision encoder, projected image-token handoff, stage registry wiring, and TPU
validation are still required.

The implementation should land text first because Gemma 3 Vision checkpoints
carry a Gemma 3 text decoder under `text_config` / `language_model.*`. Once the
text path is validated, the vision PR can focus on SigLIP, projector, staged
runtime integration, and image placeholder replacement.

TPU availability for development and validation: required smokes are scoped to
`v6e-1` and `v6e-4` where possible. Larger-shape TP and long-context validation
are follow-up work as capacity allows.

## 3. Baseline Already In Repo

| Component | Status in `main` |
|---|---|
| `Gemma3ForConditionalGeneration` listed in `multimodal_model_archs` | Present in `python/sgl_jax/srt/configs/model_config.py` |
| Gemma 2 text decoder, `GemmaRMSNorm`, `RadixAttention`, `RotaryEmbedding`, `LinearBase`, `Embed` | Present (`srt/models/gemma2.py`, `srt/layers/...`) |
| Multimodal staged runtime | Present per #552 (`srt/multimodal/{manager,model_executor,models}/...`) |
| Stage config registry | Present in `srt/multimodal/models/static_configs/yaml_registry.py` |
| Stage class resolver | Present in `srt/multimodal/manager/stage.py::get_model_class()` |
| SigLIP vision encoder | Absent; this RFC adds it |
| Gemma 3 text decoder, vision stage configs, and Gemma 3 weight mappings | Absent; this RFC adds them |

## 4. Variant Coverage

Grouped by architecture, not per checkpoint. All sizes within a row land in the
same milestone unless noted.

| Family | Representative checkpoints | HF architecture | Modalities | Milestone |
|---|---|---|---|---|
| Gemma 3 core text | `google/gemma-3-270m{,-it}`, `google/gemma-3-1b-{pt,it}` | `Gemma3ForCausalLM` | Text | M1-M2 |
| Gemma 3 core vision | `google/gemma-3-{4b,12b,27b}-{pt,it}` | `Gemma3ForConditionalGeneration` | Text, image | M3-M4 |
| Gemma 3 QAT / GGUF | QAT artifacts in the Gemma 3 release collection | Same architecture where loader-compatible | Text / image | Loader-format follow-up |
| ShieldGemma 2 | `google/shieldgemma-2-4b-it` | `ShieldGemma2ForImageClassification` | Image classification | Explicit non-goal; should be rejected from Gemma generation stage routing |

### 5.1 Gemma 3 Text Stack (`Gemma3ForCausalLM`)

The closest local starting point is `python/sgl_jax/srt/models/gemma2.py`.
Gemma 3 differs from Gemma 2 in these concrete ways:

| Aspect | Gemma 2 in repo | Gemma 3 requirement |
|---|---|---|
| MLP activation | `jax.nn.gelu` | `jax.nn.gelu(approximate=True)` / `gelu_pytorch_tanh` |
| Logit softcapping | `final_logit_softcapping`, `attn_logit_softcapping` | Removed |
| Q/K norm | Absent | `GemmaRMSNorm(head_dim)` on Q and K per head before RoPE |
| Attention pattern | 1:1 sliding/full alternation | Iterate `config.layer_types[i]` rather than assume any fixed sliding:full ratio |
| RoPE parameters | Single `rope_theta` | Per layer type: sliding and full attention use distinct `rope_parameters` |
| Head dim | Typical 128 | Respect `config.head_dim`; 270M/1B use `head_dim=256` |
| Sliding-window boundary | Existing Gemma 2 behavior | HF/SGLang GPU use an inclusive window; sglang-jax `RadixAttention` is exclusive, so pass `sliding_window - 1` |

Other invariants:

- Embeddings are scaled by `sqrt(hidden_size)` after lookup.
- Final `GemmaRMSNorm` runs before the LM head.
- The standard `LogitsProcessor` produces logits.
- Tied input/output embeddings are expected for the public HF configs.
- Tensor parallel sharding can follow the Gemma 2 pattern for projections while
  the per-head q/k norm weights stay replicated.

### 5.2 Text And Conditional Weight Mappings

M1 must cover both text-only checkpoints and the language-model prefix used by
vision checkpoints.

| HF text checkpoint key | HF vision checkpoint key | sglang-jax target | Shape | Sharding expectation |
|---|---|---|---|---|
| `model.embed_tokens.weight` | `language_model.model.embed_tokens.weight` | `model.embed_tokens.embedding` or tied LM head target | `[vocab_size, hidden_size]` | Existing embedding sharding |
| `model.layers.N.self_attn.q_proj.weight` | `language_model.model.layers.N.self_attn.q_proj.weight` | `model.layers.N.self_attn.q_proj.kernel` | `[num_heads * head_dim, hidden_size]` logically | Tensor-parallel projection axis |
| `model.layers.N.self_attn.k_proj.weight` | `language_model.model.layers.N.self_attn.k_proj.weight` | `model.layers.N.self_attn.k_proj.kernel` | `[num_kv_heads * head_dim, hidden_size]` logically | Tensor-parallel projection axis, with existing KV-head replication when needed |
| `model.layers.N.self_attn.v_proj.weight` | `language_model.model.layers.N.self_attn.v_proj.weight` | `model.layers.N.self_attn.v_proj.kernel` | `[num_kv_heads * head_dim, hidden_size]` logically | Tensor-parallel projection axis, with existing KV-head replication when needed |
| `model.layers.N.self_attn.o_proj.weight` | `language_model.model.layers.N.self_attn.o_proj.weight` | `model.layers.N.self_attn.o_proj.kernel` | `[hidden_size, num_heads * head_dim]` logically | Tensor-parallel input axis |
| `model.layers.N.self_attn.q_norm.weight` | `language_model.model.layers.N.self_attn.q_norm.weight` | `model.layers.N.self_attn.q_norm.weight` | `[head_dim]` | Replicated, no tensor axis |
| `model.layers.N.self_attn.k_norm.weight` | `language_model.model.layers.N.self_attn.k_norm.weight` | `model.layers.N.self_attn.k_norm.weight` | `[head_dim]` | Replicated, no tensor axis |
| `model.layers.N.mlp.gate_proj.weight` | `language_model.model.layers.N.mlp.gate_proj.weight` | `model.layers.N.mlp.gate_proj.kernel` | MLP projection shape | Tensor-parallel projection axis |
| `model.layers.N.mlp.up_proj.weight` | `language_model.model.layers.N.mlp.up_proj.weight` | `model.layers.N.mlp.up_proj.kernel` | MLP projection shape | Tensor-parallel projection axis |
| `model.layers.N.mlp.down_proj.weight` | `language_model.model.layers.N.mlp.down_proj.weight` | `model.layers.N.mlp.down_proj.kernel` | MLP projection shape | Tensor-parallel input axis |
| `model.norm.weight` | `language_model.model.norm.weight` | `model.norm.weight` | `[hidden_size]` | Replicated |

No-download mapping tests should assert that `self_attn.q_norm.weight` and
`self_attn.k_norm.weight` are consumed for both the text-only layout and the
`language_model.*` conditional-generation layout. These keys are a load-bearing
Gemma 2 -> Gemma 3 difference and should not be covered only indirectly by a
runtime smoke.

### 5.3 Gemma 3 Vision Stack (`Gemma3ForConditionalGeneration`)

Gemma 3 Vision adds a SigLIP tower and projector before the same text stack:

```text
Gemma3ForConditionalGeneration
  vision_tower: SigLipVisionModel
  multi_modal_projector:
    AvgPool2D
    GemmaRMSNorm(vision_hidden_size)
    Linear(vision_hidden_size -> text_hidden_size)
  language_model: Gemma3ForCausalLM
```

In sglang-jax's staged runtime, `vision_tower` and `multi_modal_projector` are
owned by the ViT-stage model runner; `language_model` is owned by the AR stage.
The class hierarchy mirrors HF, while the stage ownership defines the runtime
split.

The projector should run in the ViT stage, not inside the AR decoder. The AR
stage should receive embeddings that are already projected to the Gemma text
hidden size. The AR model-level `get_input_embeddings()` path should only merge
those projected image embeddings into image placeholder positions.

This keeps the stage boundary explicit:

```text
Multimodal tokenizer
  -> input_ids: [seq_len]
  -> mm_items: image payloads with processor-normalized pixel tensors
  -> im_token_id / image_token_id

ViT stage
  -> runs SigLIP vision tower
  -> runs AvgPool/RMSNorm/Linear projector
  -> emits projected image embeddings in Gemma text hidden size

AR stage
  -> receives input_ids and projected image embeddings
  -> get_input_embeddings() replaces image placeholder token embeddings
  -> autoregressive prefill/decode proceeds with input_embedding
```

Proposed inter-stage payload contract:

| Field | Producer | Consumer | Shape | Meaning |
|---|---|---|---|---|
| `input_ids` | tokenizer | AR stage | `[seq_len]` | Tokenized prompt, including image placeholder tokens |
| `mm_items[].feature` / `Req.pixel_values` | tokenizer / global scheduler | ViT stage | Processor-normalized image tensor, initially `[num_images, 3, image_size, image_size]`; crop/pan-and-scan expansion is follow-up | Image pixels for SigLIP |
| `image_token_id` / `im_token_id` | tokenizer | ViT and AR stages | scalar int | Placeholder token id to replace |
| projected image embeddings | ViT stage | AR stage | `[total_image_tokens, hidden_size]`, where `total_image_tokens = num_images * 256` | SigLIP + projector output, already in text hidden size |
| `multimodal_embedding` / `input_embedding` handoff | ViT/global scheduler/AR scheduler | model runner | `[seq_len, hidden_size]` after placeholder merge | Full prompt embedding used for prefill |

The first implementation can materialize the final merged prompt embedding
before AR prefill, matching the current `ForwardBatch.input_embedding` path.
The invariant is that exactly 256 projected image tokens are consumed per image
placeholder span for the 4B/12B/27B Gemma 3 Vision checkpoints.

### 5.4 Vision Weight Mappings

| HF prefix | sglang-jax target | Notes |
|---|---|---|
| `vision_tower.vision_model.*` | `SigLipVisionModel` parameters | Native SigLIP tower |
| `multi_modal_projector.mm_input_projection_weight` | Projector linear kernel | Maps projector output to text hidden size |
| `multi_modal_projector.mm_soft_emb_norm.weight` | Projector `GemmaRMSNorm.weight` | Replicated norm weight |
| `language_model.model.*` | Gemma 3 text stack | Conditional-generation text prefix |
| `language_model.lm_head.weight` | Tied embedding / LM head target | Follow text tying behavior |

Internal SigLIP weight names should follow the HF `SiglipVisionModel`
reference.

## 6. Implementation Plan

The milestones are ordered by dependency. M1/M2 establish the text model and
validation base; M3/M4 add Gemma 3 Vision.

| Milestone | Scope | Files touched (new unless noted) |
|---|---|---|
| M1: Gemma 3 text decoder | `Gemma3ForCausalLM`, decoder, attention, MLP, q/k norm, weight mapping for text-only and conditional prefixes, no-download tests | `srt/models/gemma3.py`; `python/sgl_jax/test/test_gemma3_model.py` |
| M2: Text validation on TPU | Load `gemma-3-270m` and `gemma-3-1b-it`, short-prompt smoke on `v6e-1`, greedy-token or logits parity vs HF/SGLang GPU at sequence length <= 64 | Validation only |
| M3: Gemma 3 Vision 4B | SigLIP encoder, projector in ViT stage, projected image payload, placeholder merge, tokenizer/processor wiring, stage config TP=1, registry wiring, 4B image-prompt smoke | `srt/multimodal/models/gemma3/...`; `srt/multimodal/models/static_configs/gemma3_stage_config.yaml`; `srt/multimodal/models/static_configs/yaml_registry.py`; `srt/multimodal/manager/stage.py`; `srt/multimodal/manager/multimodal_tokenizer.py`; scheduler/input-embedding handoff if needed |
| M4: Vision scale-up | TP=4 stage config, 12B/27B checkpoint validation where capacity allows, prefill/decode smoke, first correctness-oriented perf numbers | `gemma3_stage_config_tp4.yaml`; minor sharding fixes as needed |

M3 must include both:

- `yaml_registry.py` entries for the public Gemma 3 Vision checkpoint names.
- `stage.py::get_model_class()` registration for the Gemma 3 vision/projector
  and AR model classes used by the stage YAML.

Adding YAML files alone is not sufficient because the multimodal global
scheduler must be able to select the YAML and resolve each `model_class`.

## 7. Validation Plan

Per-PR no-download tests:

- Registry exposure for `Gemma3ForCausalLM` and `Gemma3ForConditionalGeneration`.
- Tiny-config text model construction.
- Text and conditional checkpoint weight mapping coverage.
- Explicit q/k norm mapping coverage for `self_attn.q_norm.weight` and
  `self_attn.k_norm.weight`, including `[head_dim]` shape and replicated TP
  sharding.
- Sliding-window value passes `sliding_window - 1` into `RadixAttention`.
- Distinct RoPE bases for `sliding_attention` vs `full_attention`.
- Vision projector payload shape: `[total_image_tokens, hidden_size]`, where
  `total_image_tokens = num_images * 256`.
- Placeholder replacement preserves token count and rejects mismatched image
  token counts.
- `yaml_registry.py` resolves Gemma 3 checkpoint names.
- `stage.py::get_model_class()` resolves Gemma 3 stage model classes.

TPU smoke tests:

- M2: `gemma-3-270m` and `gemma-3-1b-it` short text prompt on `v6e-1`.
- M3: `gemma-3-4b-it` single-image prompt on `v6e-1`, asserting exactly 256
  image tokens per image.
- M4: `gemma-3-27b-it` single-image prompt on `v6e-4` with TP=4 where capacity
  allows.

Reference checks:

- Compare selected logits or greedy tokens with HF Transformers for text-only
  checkpoints.
- Compare attention masking and sliding-window behavior with SGLang GPU or
  vLLM references.
- Compare image-token counts and placeholder replacement against the HF/SGLang
  GPU processor path.

## 8. Risks And Open Questions

1. SigLIP ownership: the first PR can place Gemma 3 SigLIP under
   `srt/multimodal/models/gemma3/`. If another model introduces a shared SigLIP
   module, refactor later rather than blocking Gemma 3 on speculative shared
   infrastructure.
2. Flash attention behavior: Gemma 3's 5:1 sliding/full pattern and per-layer
   RoPE parameters need TPU validation. If the flash backend has a correctness
   issue, route the affected layers through native attention first.
3. HF gating: gated checkpoints require an HF token on validation machines.
   No-download tests should not require credentials.
4. Pan-and-scan: the first vision slice covers the default fixed-size image path.
   Pan-and-scan correctness and dynamic image-token accounting are follow-up
   work.

## 9. Non-goals

- Gemma 3n support. It is split into a separate follow-up RFC.
- Long-context or throughput tuning beyond first correctness-oriented numbers.
- New quantized or alternate loader formats such as QAT, GGUF, or LiteRT preview
  beyond what the existing sglang-jax loader already supports.
- ShieldGemma 2 or other Gemma-adjacent classifier models.
- Tunix integration, which should be handled separately on the Tunix side.

### Related resources

- #816: Gemma3-Vision feature request.
- #178: older Gemma 3 model support issue
- #552: multimodal architecture RFC.
- #909: Q2 roadmap.
- SGLang GPU: `sglang/python/sglang/srt/models/gemma3_causal.py`,
  `gemma3_mm.py`.
- vLLM GPU: `vllm/vllm/model_executor/models/gemma3.py`, `gemma3_mm.py`.
- vLLM TPU plugin / tpu-inference: `tpu-inference/tpu_inference/models/jax/`
  for JAX-on-TPU sharding and loader patterns where applicable.
- HF Transformers: `transformers/models/gemma3/`.
- SGLang Gemma 3 reference PR: <https://github.com/sgl-project/sglang/pull/4424>.
- Technical report: <https://arxiv.org/abs/2503.19786>.

**Comment by pengchengneo** [comment]
thanks for your conrtibution @hashkanna 
Overall direction looks good, especially landing Gemma3 text first and using it as the base for vision. I think the RFC needs a bit more precision in three places before implementation starts:

1.  Please specify whether AvgPool/RMSNorm/Linear runs inside the ViT stage or inside the AR `get_input_embeddings()` path, and define the payload shape between stages.
2.  Please explicitly list the Gemma3 q/k norm weight mappings (`self_attn.q_norm`, `self_attn.k_norm`) and sharding/shape expectations. These are a  key Gemma2 -> Gemma3 difference and should be covered by no-download mapping tests.
3. M3 should include stage config registry wiring (`yaml_registry.py`) and `stage.py.get_model_class()` registration, otherwise adding the YAML files classes is not enough for the multimodal global scheduler to select Gemma3.

I would also consider splitting Gemma3n into a separate RFC after core Gemma3 text+vision lands, since its architecture and tower decisions are mostly independent.

**Comment by hashkanna** [comment]
Thanks! this makes sense. I have split Gemma3n into a new RFC #1078. 
I have updated this RFC based on your comments as below.

1. The vision stage boundary is now explicit: SigLIP + AvgPool/RMSNorm/Linear projector run in the ViT stage, and the AR path only merges projected image embeddings at image placeholder positions.
2. The inter-stage payload is now defined as projected image embeddings with shape `[total_image_tokens, hidden_size]`, where `total_image_tokens = num_images * 256`.
3. The q/k norm mappings are listed explicitly for both text-only and `language_model.*` checkpoint layouts, including `[head_dim]` shape and replicated TP sharding. I also added this to the no-download test plan.
4. M3 now explicitly includes both `yaml_registry.py` entries and `stage.py::get_model_class()` registration, with validation coverage for both.

**Comment by pengchengneo** [comment]
Okay, please start development. 

Also, we hope you'll submit a small pull request (PR) after completing a section, allowing for gradual merging and facilitating review and issue identification. @hashkanna

**Comment by hashkanna** [comment]
sounds good. Thanks!

**Comment by hashkanna** [comment]
@pengchengneo can you review this PR #1100  thank you!
